File: go/notifications.go
package nitroping

import (
	"context"
	"errors"
	"net/url"
)

// NotificationsService wraps the /api/v1/notifications endpoints.
// Construct one via NewClient — there is no value in using this type
// directly.
type NotificationsService struct {
	transport *transport
}

// Send enqueues a notification. On success the result is decoded into
// a NotificationResult (typically {ID, Status: "queued"}).
//
// Per-call options (currently only WithIdempotencyKey, WithHeader) are
// passed as trailing variadic args. The Idempotency-Key + body
// combination is honoured server-side: same key + same body replays
// the cached response, same key + different body returns a 409 with
// code "idempotency_conflict".
//
//	res, err := client.Notifications.Send(ctx, nitroping.SendRequest{
//	    Title:  "Order #4129 shipped",
//	    Body:   "On its way",
//	    Target: nitroping.AllDevices(),
//	}, nitroping.WithIdempotencyKey("order-shipped-4129"))
func (s *NotificationsService) Send(
	ctx context.Context,
	req SendRequest,
	opts ...RequestOption,
) (*NotificationResult, error) {
	cfg := applyOptions(opts)
	var out NotificationResult
	if err := s.transport.do(ctx, "POST", "/api/v1/notifications", req, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get fetches a previously-enqueued notification by id. The full row
// (including the `counters` map) is returned as a generic map so the
// SDK doesn't have to track every server-side schema change.
//
// To read a typed field, key into the map:
//
//	row, err := client.Notifications.Get(ctx, "abc-123")
//	if err != nil { return err }
//	counters, _ := row["counters"].(map[string]any)
func (s *NotificationsService) Get(
	ctx context.Context,
	id string,
	opts ...RequestOption,
) (map[string]any, error) {
	if id == "" {
		return nil, errors.New("nitroping: notification id is required")
	}
	cfg := applyOptions(opts)
	var out map[string]any
	path := "/api/v1/notifications/" + url.PathEscape(id)
	if err := s.transport.do(ctx, "GET", path, nil, cfg, &out); err != nil {
		return nil, err
	}
	return out, nil
}

File: js/src/notifications.ts
/**
 * `notifications` resource client.
 *
 * Mounted on `Nitroping` as `np.notifications`. Wraps
 * `POST /api/v1/notifications` and `GET /api/v1/notifications/:id`.
 */

import type { HttpClient } from "./http"
import type { NotificationResponse, SendNotificationRequest } from "./types"

/** Per-call overrides for `send()`. */
export interface SendOptions {
  /**
   * Optional `Idempotency-Key`. If the same key + the same body is sent
   * again the server replays the cached response. Same key + different
   * body returns a 409 (`idempotency_conflict`).
   *
   * Max 255 characters. Pick something stable + unique for the logical
   * operation (e.g. `order-shipped-4129`).
   */
  idempotencyKey?: string
}

export class NotificationsClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Enqueue a new notification.
   *
   * Returns `{ id, status }` on `201 Created`. On non-2xx the SDK
   * throws a `NitropingError` carrying the server's `code`, `message`,
   * and (for validation failures) the per-field `details` map.
   */
  async send(
    input: SendNotificationRequest,
    options: SendOptions = {},
  ): Promise<NotificationResponse> {
    const headers: Record<string, string> = {}
    if (options.idempotencyKey !== undefined) {
      headers["Idempotency-Key"] = options.idempotencyKey
    }

    return await this.http.request<NotificationResponse>("POST", "/api/v1/notifications", {
      body: toWire(input),
      headers,
    })
  }

  /** Fetch a previously enqueued notification by id. */
  async get(id: string): Promise<Record<string, unknown>> {
    return await this.http.request("GET", `/api/v1/notifications/${encodeURIComponent(id)}`)
  }

  /**
   * Cancel a scheduled or in-flight notification. Wraps
   * `DELETE /api/v1/notifications/:id`.
   *
   * Returns `{ id, status: "canceled" }`. Throws a `NitropingError`
   * with `code: "cannot_cancel"` (409) if the notification already
   * reached a terminal state, or `code: "not_found"` (404).
   */
  async cancel(id: string): Promise<{ id: string; status: string }> {
    return await this.http.request("DELETE", `/api/v1/notifications/${encodeURIComponent(id)}`)
  }
}

/**
 * Convert the camelCase SDK shape into the snake_case wire shape the
 * Phoenix controller expects.
 */
function toWire(input: SendNotificationRequest): Record<string, unknown> {
  const wire: Record<string, unknown> = {}
  if (input.title !== undefined) wire["title"] = input.title
  if (input.body !== undefined) wire["body"] = input.body
  if (input.template !== undefined) wire["template"] = input.template
  if (input.vars !== undefined) wire["vars"] = input.vars
  if (input.data !== undefined) wire["data"] = input.data
  if (input.icon !== undefined) wire["icon"] = input.icon
  if (input.image !== undefined) wire["image"] = input.image
  if (input.clickAction !== undefined) wire["click_action"] = input.clickAction
  if (input.deepLink !== undefined) wire["deep_link"] = input.deepLink
  if (input.actions !== undefined) wire["actions"] = input.actions
  if (input.scheduledAt !== undefined) wire["scheduled_at"] = input.scheduledAt
  if (input.expiresAt !== undefined) wire["expires_at"] = input.expiresAt
  wire["target"] = targetToWire(input.target)
  return wire
}

function targetToWire(target: SendNotificationRequest["target"]): Record<string, unknown> {
  if ("all" in target) return { all: target.all }
  if ("deviceIds" in target) return { device_ids: target.deviceIds }
  if ("userIds" in target) return { user_ids: target.userIds }
  if ("tags" in target) return { tags: target.tags }
  // `target` is a discriminated union — exhaustiveness is enforced at
  // the type level, but fall back to passing through for forward compat.
  return target as Record<string, unknown>
}
